### ISSUE TITLE
How to build stacked Sequence-to-sequence autoencoder?   

### ISSUE THREAD

**Issue by DeepWolf90** [keras-team/keras#9093]
Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by DeepWolf90** [comment]
in keras blog:"Building Autoencoders in Keras"
the following code is provided to build single sequence to sequence autoencoder

from keras.layers import Input, LSTM, RepeatVector
from keras.models import Model

inputs = Input(shape=(timesteps, input_dim))
encoded = LSTM(latent_dim)(inputs)

decoded = RepeatVector(timesteps)(encoded)
decoded = LSTM(input_dim, return_sequences=True)(decoded)

sequence_autoencoder = Model(inputs, decoded)
encoder = Model(inputs, encoded)

I want to build stacked autoencoder, how to update this code to build stacked autoencoder?

**Comment by DeepWolf90** [comment]
i try it myself, and this is my code:

timesteps = 3
input_dim = 1
inputs = Input(shape=(timesteps, input_dim))
encoded = LSTM(4)(inputs)
encoded = RepeatVector(timesteps)(encoded)
encoded = LSTM(2)(encoded)
encoded = RepeatVector(timesteps)(encoded)
decoded =  LSTM(4,return_sequences = True)(encoded)
decoded =  LSTM(input_dim,return_sequences = True)(decoded)

sequence_autoencoder = Model(inputs, decoded)

sequence_autoencoder.compile(loss='mean_squared_error', optimizer='Adam')

sequence_autoencoder.fit(x_train, x_train,
                         epochs=100,
                         batch_size=1,
                         shuffle=True,
                         )

i want to know, is this code correct or am i missing something?

**Comment by visionscaper** [comment]
Use `return_sequences=True` for all encoder layers, except for the last, than you can ditch the in-between `RepeatVector` layers. Only after the last encoder layer you will need the `RepeatVector` layer. The same holds for the decoder layers. Obviously you need the `return_sequences=True` for your last decoder layer, as you already have.

**Comment by DeepWolf90** [comment]
thank you for your help.
theoretically, both models(mine and your suggestion) are the same, but they give different result.
i tried both for the same architecture and database, then evaluated them, mine gives 0.01602 and yours gives 0.01859 in RMSE. is there a reason for this?

**Comment by visionscaper** [comment]
Hi @DeepWolf90,

I'm not sure, can you share the code for both models you trained?

Further, what activation function did you use for the RNN outputs?
Without any specification a linear activation is used.
I advice you to read the Class documentation for the classes (e.g. `LSTM`) you use:

    class LSTM(RNN):
        """Long-Short Term Memory layer - Hochreiter 1997.

        # Arguments
            units: Positive integer, dimensionality of the output space.
            activation: Activation function to use
                (see [activations](../activations.md)).
                If you pass None, no activation is applied
                (ie. "linear" activation: `a(x) = x`).
            recurrent_activation: Activation function to use
                for the recurrent step
                (see [activations](../activations.md)).

**Comment by DeepWolf90** [comment]
thank you @visionscaper  for your interest. 
Regarding the both models issue, i did further research and found out that, what you suggested about Use return_sequences=True for all encoder layers, except for the last encoder is "RIGHT".
First, i trained SAE and this is the code:
    timesteps = x_train.shape[1]
    input_dim = 1
    inputs = Input(batch_shape=(1,timesteps, input_dim))
    encoded = LSTM(4,return_sequences = True)(inputs)
    encoded = LSTM(3,return_sequences = True)(encoded)
    encoded = LSTM(2)(encoded)
    decoded = RepeatVector(timesteps)(encoded) 
    decoded =  LSTM(3,return_sequences = True)(decoded)                                   
    decoded =  LSTM(4,return_sequences = True)(decoded)
    decoded =  LSTM(input_dim,return_sequences = True)(decoded)

    sequence_autoencoder = Model(inputs, decoded)

    encoder = Model(inputs,encoded)

    sequence_autoencoder.compile(loss='mean_squared_error', optimizer='Adam')

    sequence_autoencoder.fit(x_train, x_train,
                         epochs=100,
                         batch_size=1,
                         shuffle=True,
                         )

second, i try to feed the encoder's output to another LSTM layer to forecast one step ahead and this is the code:
    lstm1 =  LSTM(3,stateful = True)(encoder.outputs)
    dropout = Dropout(0.3)(lstm1)
    out = Dense(1)(dropout)
    model = Model(encoder.inputs,out)
    model.compile(loss='mean_squared_error', optimizer='Adam')

    model.fit(x_train, y_train, epochs=200, batch_size=1, verbose=1, shuffle=True)

But, it gives me error because the encoder.outputs is 2 Dim and LSTM needs 3 Dim.

**Comment by nis192** [comment]
Did you find out the problem?

**Comment by mg64ve** [comment]
I think you should put return_sequences = False before RepeatVector and use TimeDistributed as final layer.
What about of making this model stateful ?

**Comment by ramesh720** [comment]
> thank you @visionscaper for your interest.
> Regarding the both models issue, i did further research and found out that, what you suggested about Use return_sequences=True for all encoder layers, except for the last encoder is "RIGHT".
> First, i trained SAE and this is the code:
> 
> ```
> timesteps = x_train.shape[1]
> input_dim = 1
> inputs = Input(batch_shape=(1,timesteps, input_dim))
> encoded = LSTM(4,return_sequences = True)(inputs)
> encoded = LSTM(3,return_sequences = True)(encoded)
> encoded = LSTM(2)(encoded)
> decoded = RepeatVector(timesteps)(encoded) 
> decoded =  LSTM(3,return_sequences = True)(decoded)                                   
> decoded =  LSTM(4,return_sequences = True)(decoded)
> decoded =  LSTM(input_dim,return_sequences = True)(decoded)
> 
> sequence_autoencoder = Model(inputs, decoded)
> 
> encoder = Model(inputs,encoded)
> 
> sequence_autoencoder.compile(loss='mean_squared_error', optimizer='Adam')
> 
> sequence_autoencoder.fit(x_train, x_train,
>                      epochs=100,
>                      batch_size=1,
>                      shuffle=True,
>                      )
> ```
> 
> second, i try to feed the encoder's output to another LSTM layer to forecast one step ahead and this is the code:
> 
> ```
>  lstm1 =  LSTM(3,stateful = True)(encoder.outputs)
>  dropout = Dropout(0.3)(lstm1)
>  out = Dense(1)(dropout)
>  model = Model(encoder.inputs,out)
>  model.compile(loss='mean_squared_error', optimizer='Adam')
> 
>  model.fit(x_train, y_train, epochs=200, batch_size=1, verbose=1, shuffle=True)
> ```
> 
> But, it gives me error because the encoder.outputs is 2 Dim and LSTM needs 3 Dim.

sir how the shape of x_train looks like
